core/src/main/scala/org/apache/spark/eventhubs/PartitionsStatusTracker.scala

```diff
@@ -0,0 +1,379 @@
+import org.apache.spark.eventhubs.utils.ThrottlingStatusPlugin
+import org.apache.spark.internal.Logging
+
+
+class PartitionsStatusTracker extends Logging{
+
+  import PartitionsStatusTracker._
+
+  // retrives the batchStatus object based on the local batchId
+  private val batchesStatus = mutable.Map[Long, BatchStatus]()
+
+  /**
+   * retirives the local batchId for a pair of (NameAndPartition, SequenceNumber)
+   * it's useful to accss the right batch when a performance metric message is
+   * received for a parition-RequestSeqNo pair.
+   * it's getting updated every time a batch is removed or added to the tracker
+   */
+  private val partitionSeqNoPairToBatchIdMap = mutable.Map[String, Long]()
+
+  /**
+   * Add a batch to the tracker by creating a BatchStatus object and adding it to the map.
+   * Also, we add mappings from each partition-startSeqNo pair to the batchId in order to be able to retrive
+   * the batchStatus object from the map when the performance metric message is received.
+   * Not that we ignore partitions with batchSize zero (startSeqNo == latestSeqNo) since we won't recieve
+   * any performance metric message for such partitions.
+   */
+  def addBatch(batchId: Long,
+               offsetRanges: Array[OffsetRange]): Unit = {
+    if(batchesStatus.contains(batchId)) {
+      throw new IllegalStateException(
+        s"Batch with local batch id = $batchId already exists in the partition status tracker. Can't be added again.")
+    }
+    // find partitions with a zero size batch.. No performance metric msg will be received for those partitions
+    val isZeroSizeBatchPartition: Map[NameAndPartition, Boolean] =
+      offsetRanges.map(range  => (range.nameAndPartition, (range.fromSeqNo == range.untilSeqNo)))(breakOut)
+
+    // create the batchStatus tracker and add it to the map
+    batchesStatus(batchId) = new BatchStatus(batchId, offsetRanges.map(range => {
+      val np = range.nameAndPartition
+      (np, new PartitionStatus(np, range.fromSeqNo, isZeroSizeBatchPartition(np)))
+    })(breakOut))
+
+    // add the mapping from partition-startSeqNo pair to the batchId ... ignore partitions with zero batch size
+    offsetRanges.filter(r => !isZeroSizeBatchPartition(r.nameAndPartition)).foreach(range => {
+      val key = partitionSeqNoKey(range.nameAndPartition, range.fromSeqNo)
+      addPartitionSeqNoToBatchIdMapping(key, batchId)
+    })
+  }
+
+  /**
+   * Remove a batch from the tracker by finding and deleting the batchStatus object from the map.
+   * Also, we remove mappings from each partition-startSeqNo pair to the batchId since this batch is old
+   * and should not be updated even if we receive a performacne metric message for it later.
+   * Not that we ignore partitions with batchSize zero (PartitionStatus.emptyBatch) since those haven't been
+   * added to the mapping when we added the batch to the tracker. [See addBatch]
+   */
+  def removeBatch(batchId: Long): Unit = {
```

Comment:
What about creating a simple circular buffer with fixed size and keep the status of three batches and allow update API only instead of doing add/remote and ordering the result which is unnecessarily complicated in terms of API usage?

Reply:
Modified the code